### PR TITLE
[DOC] - Tutorial for morlet wavelets

### PR DESCRIPTION
**Pull request by agquinn**
This is a tutorial for performing time-frequency decomposition using morlet wavelets following the neurodsp.timefrequency.wavelets module.

### CODE REVIEW

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:7]: Maybe "explore" is a better choice of words about a module rather than "follow"

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:72]: `filter_kwargs` is not defined here, it needs to be defined before something like `filter_kwards=dict(l_freq=5)` but probably not with that argument

Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:72]: Also it looks like you have a hanging parentheses at the end and you can either define a dictionary like this `components=dict(sim_powerlaw=dict(exponent=-2))` or like this `components={'sim_powerlaw': {'exponent': -2}}`. I prefer the first for user-facing code such as a tutorial but it should be consistent with neurodsp style so check to see what other tutorials use. You're mixing the two with `{exponent=-2}` which doesn't work

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:85]: Usually you put the section header and explanation above the plots

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:115]: You can't have a positional argument (without an equals sign) after a keyword argument (`fs=500`) it's just not allowed in python. Best practice, again for user-facing code is to use keywords for everything unless it's totally obvious like `sig` (many times the first argument is totally obvious)

Comment (alexrockhill) [reply tutorials/timefreq/wavelets_tutorial.py:115]: `mwt = compute_wavelet_transform(sig, fs=500, freqs=freqs)` computes for me

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:121]: This should probably go at the top of the script with the other imports

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:122]: This doesn't compute because you have imaginary values. Usually people look at the real values (power) but the imaginary values tell phase that can also be useful. Might be something worth adding but for now, this makes the plot I think you're trying to make `plt.imshow(abs(mwt), aspect='auto')`

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:159]: You never use `cts`

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:165]: You never define `mwt_real` if this worked for you, you probably did some computations that you didn't save in the code

**Comment by alexrockhill** [comment]
To provide a general comment about the goal of the tutorial, I think it might help to think about what someone who wants to use the `neurodsp` morlet wavelets functionality would want to get out of a tutorial. I think they would want to know:

- How to use the functions
- That they work by recapitulating simulated data
- How and when to change the parameters

The first two goals are pretty close but I think need more explaining and I think the third could really use a lot of explaining and some plots.

@agquinn, when you open a PR you want to reference the issue that you are addressing like so:
"addresses https://github.com/neurodsp-tools/neurodsp/issues/166"

@TomDonoghue et al. @agquinn and I are working on this tutorial together as a bit of a learning exercise so feel free to review whenever or at the end I can tag you to let you know my review is done.

**Comment by alexrockhill** [comment]
Also morlet wavelets are a time-frequency method and the plots I saw from running the tutorial where homogenous in time so that is something that could due to be changed as well

**Comment by ryanhammonds** [comment]
Hi @alexrockhill and @agquinn, thanks for the PR! Feel free to tag myself (and Tom) for a review when ready.

File: tutorials/timefreq/wavelets_tutorial.py
Comment (alexrockhill) [inline tutorials/timefreq/wavelets_tutorial.py:98]: This line is way, way too long and it doesn't run. The maintainers of NeuroDSP have chosen to make lines have to be less than 100 characters, usually it's less than 80 in python but it doesn't make a huge difference whether it's 80 or 100, it's really about all the text fitting on an average screen.

I'm going to stop reviewing now because of the syntax error on this line so if you could write a comment in the conversations tab and tag @alexrockhill when you're ready for a review that would be great.

**Comment by alexrockhill** [comment]
It's a bit of a sphinx idiosyncrasy but in order for a tutorial to plot, the file name must begin with `plot_` so it should be changed to `plot_wavelet_tutorial.py`

**Comment by alexrockhill** [comment]
And probably wavelet and not wavelets in the name

**Comment by TomDonoghue** [comment]
Hey @agquinn and @alexrockhill - this is super cool that you are doing this, a big thank you for choosing to contribute to NDSP! Adding a wavelet tutorial sounds great to me!

Please feel free to tag myself or Ryan in if you need help with anything with the module, or if it's ready for a general review. 

I've had a quick scan - for some overview notes, I'd just add that you can check the other tutorials to match conventions / layouts and that there might want to be more code / examples / plots (as compared to the amount of text), including having time varying oscillations (using `sim_bursty_oscillation`, perhaps), and exploring different settings, etc.

File: tutorials/timefreq/wavelets_tutorial.py
Comment (agquinn) [reply tutorials/timefreq/wavelets_tutorial.py:98]: I'm kind of struggling with this line. I can't figure out how to add the filter key word arguments, and I don't know how to make the dictionary either. It's hard for me to read using the curly bracket notation, but I know that's how it looks in the other Neurodsp tutorials, so I didn't think I should change it. I thought maybe I should add a line above for the filter_signal to define the key word arguments, like this: `filter_signal(sig=sig, fs=fs, pass_type=lowpass,  f_range=10)
`
But I don't know where to add this line, since sig also relies on what gets defined in the components

**Comment by agquinn** [comment]
@alexrockhill I know I haven't resolved all of the issues yet, but I'm kind of stuck on how to plot the Morlet wavelet transformed series. I can't figure out why they don't look like individual bursts, and I've tried changing the frequency parameters, and messing with the normalization methods, but I'm still kind of stuck on this part. I don't know if you have any advice/ideas for this part, but I'd appreciate the help.

**Comment by alexrockhill** [comment]
> @alexrockhill I know I haven't resolved all of the issues yet, but I'm kind of stuck on how to plot the Morlet wavelet transformed series. I can't figure out why they don't look like individual bursts, and I've tried changing the frequency parameters, and messing with the normalization methods, but I'm still kind of stuck on this part. I don't know if you have any advice/ideas for this part, but I'd appreciate the help.

So it's hard to review because the python file is called `wavelets_tutorial.py` and not `plot_wavelets_tutorial.py` so it doesn't run in the documentation... I know you want feedback but you really have to incorporate the feedback already given first.

**Comment by alexrockhill** [comment]
It's also hard to review because you have a lot of code written and especially a ton of comments that in total is way too much. You can save the extra code somewhere outside the project directory (or on a different branch i.e. `git checkout -b tmp` and then `git checkout wavelet_tutorial` and then delete most of the wavelets_tutorial.py file but just save the good parts and then if you ever need anything you deleted you can do `git checkout tmp` and the file will be how you left it and you can copy it and `git checkout wavelet_tutorial` and paste it back in).

As for the core of the code here is mostly what you wrote with a few changes:

```
import numpy as np
import matplotlib.pyplot as plt
from neurodsp.sim import sim_bursty_oscillation
from neurodsp.utils import set_random_seed, create_times
from neurodsp.plts.time_series import plot_time_series
from neurodsp.timefrequency.wavelets import compute_wavelet_transform

set_random_seed(0)

fs = 500
n_seconds = 10
freq = 20

# Simulate a signal with bursty oscillations at 20 Hz with a decay time of 2 seconds
sig = sim_bursty_oscillation(n_seconds, fs, freq)
times = create_times(n_seconds, fs)

freqs = np.linspace(5, 100, 50)
mwt = compute_wavelet_transform(sig, fs=fs, n_cycles=7, freqs=freqs)

fig, ax = plt.subplots()
ax.imshow(abs(mwt), aspect='auto')
ax.invert_yaxis()
ax.set_xlabel('time (s)')
ax.set_xticks(np.linspace(0, times.size, 5))
ax.set_xticklabels(np.round(np.linspace(times[0], times[-1], 5), 2))
ax.set_ylabel('freq (Hz)')
ax.set_yticks(np.linspace(0, freqs.size, 5))
ax.set_yticklabels(np.round(np.linspace(freqs[0], freqs[-1], 5), 2))
fig.show()
```

Going forward, I would suggest deleting everything you have and starting with this code and adding comments to explain it. You just have so much going on in what you started that it's not very helpful for someone trying to learn about the Morlet wavelet convolution method. I would try thinking about what would you want to read if you had never heard of this so that you could code it up in python, use it and understand it. You shouldn't just copy it from here https://mne.tools/stable/auto_examples/time_frequency/plot_time_frequency_simulated.html?highlight=morlet but most of the information is already described elsewhere, the point is to make it easy for new neuroDSP users to use and understand Morlet wavelets.

**Comment by alexrockhill** [comment]
That's a good start! Now it needs a couple of sentences explaining what's going on at the top and then maybe going into what's going on under the hood by making the wavelet and plotting it as you were trying to do at the end of the very long version that you had previously.

Also "Wavelets" is not as informative as "Morlet Wavelet Analysis" as a title.

**Comment by agquinn** [comment]
@alexrockhill I think I've resolved most of the comments that came up earlier.

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (alexrockhill) [inline tutorials/timefreq/plot_MorletWavelet.py:45]: Are these too long? It looks weird with the `#`s being much shorter than the text, usually the text stays within them but I believe the lines just have to be under 100 characters

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (alexrockhill) [inline tutorials/timefreq/plot_MorletWavelet.py:71]: You imported a `create_freqs` function, you'd probably want to use it here if you're going to use it.

If not, you should delete the import as it is "cruft" or extraneous code. In that vein, you should delete all the function imports that you don't use.

**Comment by alexrockhill** [comment]
This is a great start, not too text heavy now, I can read it easily and follow along. I made a few minor comments about the style etc. that would be great if you could fix.

The main thing that would be good to add, is that although this tutorial tells you how to use the Morlet Wavelet transform, it tells you nothing about what's happening under the hood. Quoting from here https://mne.tools/dev/auto_examples/time_frequency/plot_time_frequency_simulated.html "[Morlet Wavelets] are a sinusoidal wave with a gaussian envelope. We can control the balance between spectral and temporal resolution with the `n_cycles` parameter, which defines the number of cycles to include in the window." This is a good start but that's from a tutorial that compares a bunch of time frequency methods. When just describing the Morlet method it would be good to go into what exactly is happening. I would add some information from other sources perhaps starting with the wiki and finding an article to cite (https://en.wikipedia.org/wiki/Morlet_wavelet) (you'll want to look and see how the other tutorials cite articles. The amount of depth you want to go into it is up to you, maybe just adding some similar plots to the wiki at a minimum and reading and really going into plotting and visualizing if you have the time, but something under the hood would be helpful to users.

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (TomDonoghue) [reply tutorials/timefreq/plot_MorletWavelet.py:45]: quick note since I noticed this comment: 
In terms of rendering the website, it doesn't really matter how long the text lines are. By convention, and for readability of the source files, we do tend to keep text lines under 100 characters. Notably, as long as you continue on the next line, a line break in the file doesn't enforce a line break on the render, so adding line breaks doesn't make the sphinx page go weird (you can compare the layout of existing tutorial files, and what they look like on the website). 
tl-dr: best practice is to keep text lines < 100 chars.

Also, in line 42/43 (and any similar cases): lengths of the title underhang (---) should match length of the title

Comment (alexrockhill) [reply tutorials/timefreq/plot_MorletWavelet.py:45]: Thanks!

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (alexrockhill) [inline tutorials/timefreq/plot_MorletWavelet.py:137]: You're speaking on behalf of all the `neurodsp` contributors, let's not use "I"

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (alexrockhill) [inline tutorials/timefreq/plot_MorletWavelet.py:128]: I think the order of these is a bit off, I would first explain the parameters and how to choose different ones and then explain how the Morlet wavelets work like this section.

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (alexrockhill) [inline tutorials/timefreq/plot_MorletWavelet.py:158]: Like Tom said, it would be nice to keep these from going over the `#`s.

**Comment by alexrockhill** [comment]
Also it looks like someone has already added a tutorial on wavelets...

If you want to add to it you can do

```
git checkout master
git pull upstream master
git checkout wavelet_tutorial
git rebase -i master
```

Then look at the file and fix all the conflicts between their tutorial and yours they will be separated by a line of  `=`s, then do

```
git add -A
git rebase --continue
```

Then the next push will have to be `git push origin wavelet_tutorials --force`

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (ryanhammonds) [inline tutorials/timefreq/plot_MorletWavelet.py:221]: range -> resolution

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (ryanhammonds) [inline tutorials/timefreq/plot_MorletWavelet.py:158]: I think `compute_wavelet_transform` could still be used here since it uses scipy's morlet function and returns a complex array.

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (ryanhammonds) [inline tutorials/timefreq/plot_MorletWavelet.py:164]: Axes could be labeled here. It may also be helpful to remind readers about convolution somewhere.

**Review by ryanhammonds — COMMENTED** [review]
Hi @agquinn, thanks for the PR! I think this tutorial will be super helpful for those who are new to wavelets - myself included. I went through the tutorial and left additional feedback/thoughts. Feel free to tag me in when ready, and I can directly push final changes before merging in.

File: tutorials/timefreq/plot_MorletWavelet.py
Comment (TomDonoghue) [reply tutorials/timefreq/plot_MorletWavelet.py:159]: also, `fig.show()` isn't needed

**Comment by TomDonoghue** [comment]
@alexrockhill - I'm not sure about you're previous comment: there is no current wavelet tutorial? I don't think anything should need rebasing here. 

@ryanhammonds - thanks for reviewing! Let me know if anything needs me to look over it.

**Comment by agquinn** [comment]
@ryanhammonds @TomDonoghue Thank you both for looking over this! I made most of the changes now, and I think it looks okay. I wasn't sure how much detail to go into, especially for the convolution part, but I think it seems pretty easy to follow. Let me know if there is anything else I should add/change!

**Comment by agquinn** [comment]
@alexrockhill I think this looks pretty good, let me know if you have any suggestions on changing/modifying anything too!

**Comment by alexrockhill** [comment]
> @alexrockhill - I'm not sure about you're previous comment: there is no current wavelet tutorial? I don't think anything should need rebasing here.
> 
> @ryanhammonds - thanks for reviewing! Let me know if anything needs me to look over it.

@TomDonoghue Audrey and I had a Zoom call and I usually never modify the master but I wasn't thinking about it at the time and then when I build the documentation the first thing I jumped to was someone already wrote a tutorial hahaha. Oops!

**Comment by alexrockhill** [comment]
[Morlet Wavelet Analysis — neurodsp 2.2.0-dev documentation.pdf](https://github.com/neurodsp-tools/neurodsp/files/5850955/Morlet.Wavelet.Analysis.neurodsp.2.2.0-dev.documentation.pdf)

Looks like there is an error...

**Comment by agquinn** [comment]
Yeah, I'm not sure why that is...I'll look into it

**Comment by TomDonoghue** [comment]
looks like this is super close! @ryanhammonds - could you do a sweep, after your prior review, and check for any remaining suggestions, and resolve any open comments above that have been addressed. Once your happy with it, tag me for review, and I'll have a final read / check and merge when it's ready.

**Comment by ryanhammonds** [comment]
Sorry about that, I left a bad review comment that lead to the error. I pushed a fix along with a few formatting changes.

**Comment by alexrockhill** [comment]
Ready to merge?

**Review by TomDonoghue — APPROVED** [review]

**Comment by TomDonoghue** [comment]
Hey all, sorry for a massive delay on my part here - I've been moving job & city etc and this fell through the cracks. 

@agquinn & @alexrockhill : a huge thank you for the submission here to improve our documentation, we greatly appreciate it! I did a full sweep of the tutorial, and because it had been so long and I didn't want to revive this with a ToDo list since y'all have probably moved on, I made direct edits to tweak some things, and add a bit of information. I'm really liking where this tutorial is now. If you have any comments or questions on the updated version, please let us know, but otherwise I don't think anything else is needed, and this should be ready to merge. 

@ryanhammonds - I think this should be good now. Can you do a quick sweep and check I didn't introduce any weird typos / bugs, and then merge if it looks good to you?

**Pull request merged**
